From fc851a27c27227e1b178d1ebc92eac7884f67132 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:04:22 -0400
Subject: [net] usb/cdc-acm: add TIOCGICOUNT

Message-id: <1410887463-20674-191-git-send-email-dzickus@redhat.com>
Patchwork-id: 95199
O-Subject: [RHEL7 PATCH 190/591] cdc-acm: add TIOCGICOUNT
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 797ef13716ce05d26c5be12893ca75b2b003f25f
Author: Oliver Neukum <oneukum@suse.de>
Date:   Wed Nov 20 11:35:35 2013 +0100

    cdc-acm: add TIOCGICOUNT

    Simple straightforward implementation. Just returning the statistics
    gathered for TIOCMIWAIT

    Signed-off-by: Oliver Neukum <oneukum@suse.de>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/class/cdc-acm.c b/drivers/usb/class/cdc-acm.c
index 944b348..3723faf 100644
--- a/drivers/usb/class/cdc-acm.c
+++ b/drivers/usb/class/cdc-acm.c
@@ -853,6 +853,27 @@ static int wait_serial_change(struct acm *acm, unsigned long arg)
  return rv;
 }
 
+static int get_serial_usage(struct acm *acm,
+       struct serial_icounter_struct __user *count)
+{
+ struct serial_icounter_struct icount;
+ int rv = 0;
+
+ memset(&icount, 0, sizeof(icount));
+ icount.dsr = acm->iocount.dsr;
+ icount.rng = acm->iocount.rng;
+ icount.dcd = acm->iocount.dcd;
+ icount.frame = acm->iocount.frame;
+ icount.overrun = acm->iocount.overrun;
+ icount.parity = acm->iocount.parity;
+ icount.brk = acm->iocount.brk;
+
+ if (copy_to_user(count, &icount, sizeof(icount)) > 0)
+  rv = -EFAULT;
+
+ return rv;
+}
+
 static int acm_tty_ioctl(struct tty_struct *tty,
      unsigned int cmd, unsigned long arg)
 {
@@ -869,6 +890,9 @@ static int acm_tty_ioctl(struct tty_struct *tty,
  case TIOCMIWAIT:
   rv = wait_serial_change(acm, arg);
   break;
+ case TIOCGICOUNT:
+  rv = get_serial_usage(acm, (struct serial_icounter_struct __user *) arg);
+  break;
  }
 
  return rv;
-- 
1.7.1